### DISCUSSION TITLE
Use localStorage for a front-end dynamic mocking. What about back-end?

### DISCUSSION THREAD

**Discussion by baptisteArno** [mswjs/msw#1116] [Q&A] [answered]
I'm running Playwright tests on a Next.js application that has mocked authentication endpoints.

Users on my app can have different user roles (paid or free for example). I'd like to be able to dynamically switch from one role to another.

✅ On the front-end, easy, I can read a variable from `sessionStorage` or `localStorage`.
❌ But it seems on the back-end, I can't really do this, I'm obligated to return a static response. I can't find a way to tell the server to mock a particular response on runtime.

Any ideas?

**Comment by kettanaito** [comment, marked as answer]
Hey, @baptisteArno.

Controlling dynamic behaviors on the client is indeed easier as your JavaScript code (where MSW also lives) gets re-evaluated on each page load. That is not, however, the case for Node.js. I believe NextJS will execute `_app.js` _once_ upon initial load of their server, and won't re-evaluate it anymore, regardless if there are hot updates coming from the app module. I believe there's an open issue for this as well in the NextJS repo. 

I had a discussion about this with one of the NextJS community members and we haven't arrived at a good way to tackle this at the moment. 

I may recommend handling different roles on the client (if that's possible), since it's often the client who requests user/session-related info.

```js
// There's a route on the server that handles user roles,
// and we, ideally, want to mock its interactions with other services
// (i.e. a database) instead of directly mocking this client-side response.
rest.get('/api/user', (req, res, ctx) => {
  const paymentPlan = req.url.searchParams.get('role')
  
  if (paymentPlan === 'paid') {
    return res(ctx.json({ paidPlanUser }))
  }

  return res(ctx.json({ freePlanUser }))
})
```

This way you'd be assuming the server's role resolution as a given, bypassing this logic on the server. I know this may not be what you want but I'm afraid I don't have a better suggestion at the moment. 

I think once NextJS supports live reloading of their server this issue will be obsolete. NextJS server depends on your mocks if you follow MSW setup as shown in the [NextJS + MSW example](https://github.com/vercel/next.js/tree/canary/examples/with-msw). But despite that dependency, the server doesn't get refreshed upon changes to the mocks.

**Reply by baptisteArno** [reply]
Thank you for the info thorough answer 👍

**Reply by kettanaito** [reply]
My pleasure. Just keep in mind that your request handlers are still a client-side code (when speaking of the browser). So you can access `location` and other things there to build the logic your use case needs. Also, be mindful that if you're referencing browser globals you'd have to polyfill them in Node.js when reusing the same handlers for integration tests (this is often done by JSDOM).

If you find my response helpful, consider [supporting the Mock Service Worker organization](https://github.com/sponsors/mswjs). The scale of the issues and questions we get is much higher than the number of people actively maintaining the project. If you wish for MSW to continue and improve, I'd appreciate your financial input, no matter how big. Thanks!

**Comment by Bram-Zijp** [comment]
You could mock local storage/session storage on the server side. I used this for automated tests in dynamic-msw I am currently working on: https://github.com/dynamicmsw/dynamic-msw/blob/main/libs/core/setup-tests.ts

Take note that you don't want any checks for the session/local storage to be undefined or not to decide on Node.js vs browser side. Probably best to only include this for unit tests (non e2e/integration tests that run in a browser). Perhaps dynamic-msw will be the answer when it's done in about a week! Let me know more about your use case so I can perhaps alter the implementation of dynamic-msw, I am curious.
